SingularityService/src/main/java/com/hubspot/singularity/resources/TaskResource.java

```diff
@@ -322,6 +323,27 @@ public SingularityTaskCleanup killTask(@PathParam("taskId") String taskId, Optio
   }
 
   @POST
+  @Path("/task/{taskId}/metadata")
+  @ApiOperation(value="Post metadata about a task that will be persisted along with it and displayed in the UI")
+  @ApiResponses({
+    @ApiResponse(code=400, message="Invalid metadata object"),
+    @ApiResponse(code=404, message="Task doesn't exist"),
+    @ApiResponse(code=409, message="Metadata with this type/timestamp already existed")
+  })
+  @Consumes({ MediaType.APPLICATION_JSON })
+  public void postTaskMetadata(@PathParam("taskId") String taskId, final SingularityTaskMetadata taskMetadata) {
+    SingularityTaskId taskIdObj = getTaskIdFromStr(taskId);
+
+    authorizationHelper.checkForAuthorizationByTaskId(taskId, user, SingularityAuthorizationScope.WRITE);
+
+    WebExceptions.checkBadRequest(taskMetadata.getTaskId().equals(taskIdObj), "Task metadata taskId %s didn't match API path %s", taskMetadata.getTaskId(), taskId);
+
+    SingularityCreateResult result = taskManager.saveTaskMetadata(taskMetadata);
```

Comment:
If the task has already been persisted to history and removed from zk, we won't be able to update the task's metadata. Do we want to be able to?